src/drivers/hmc5883/hmc5883.cpp

```diff
@@ -32,11 +32,149 @@
  ****************************************************************************/
 
 /**
- * @file hmc5883.cpp
- *
- * Driver for the HMC5883 / HMC5983 magnetometer connected via I2C or SPI.
+ * Driver for the HMC5883 / HMC5983 magnetometer
  */
 
+#include <px4/device_driver.h>
+
+enum { CONFIG_A, CONFIG_B, MODE,
+       DATA_OUT_X_MSB, DATA_OUT_X_LSB,
+       DATA_OUT_Y_MSB, DATA_OUT_Y_LSB,
+       DATA_OUT_Z_MSB, DATA_OUT_Z_LSB,
+       STATUS, ID_A, ID_B, ID_C,
+       DATA_OUT_X, DATA_OUT_Y, DATA_OUT_Z };
+
+static const px4::reg_t regs[] = {
+	[CONFIG_A] = { 0, px4::ioperm::READWRITE, 1, 0x80 },
+	[CONFIG_B] = { 1, px4::ioperm::READWRITE, 1, 0x1f },
+	[MODE] = { 2, px4::ioperm::READWRITE, 1, 0xfc },
+
+	[DATA_OUT_X_MSB] = { 3, px4::ioperm::READ, 1, 0 },
+	[DATA_OUT_X_LSB] = { 4, px4::ioperm::READ, 1, 0 },
+	[DATA_OUT_Y_MSB] = { 7, px4::ioperm::READ, 1, 0 },
+	[DATA_OUT_Y_LSB] = { 8, px4::ioperm::READ, 1, 0 },
+	[DATA_OUT_Z_MSB] = { 5, px4::ioperm::READ, 1, 0 },
+	[DATA_OUT_Z_LSB] = { 6, px4::ioperm::READ, 1, 0 },
+
+	[STATUS] = { 9, px4::ioperm::READ, 1, 0x3},
+
+	[ID_A] = { 10, px4::ioperm::READ, 1, 0},
+	[ID_B] = { 11, px4::ioperm::READ, 1, 0},
+	[ID_C] = { 12, px4::ioperm::READ, 1, 0},
+
+	// TODO: we really need a 6-byte DATA_OUT pseudo-register,
+	// but we don't have an API for that (yet)
+	[DATA_OUT_X] = { 3, px4::ioperm::READ, 2, 0 },
+	[DATA_OUT_Y] = { 7, px4::ioperm::READ, 2, 0 },
+	[DATA_OUT_Z] = { 5, px4::ioperm::READ, 2, 0 },
+};
+
+// TODO: px4 namespace?
+
+class hmc5883 : public px4::device {
+public:
+	hmc5883(std::string &path) : device(path) {
+		io = new px4::regcache(path, regs);
+	}
+	int init();
+	// initiates periodic sampling at the specified rate
+	// call again to adjust the sampling rate
+	// hz==0 means "capture one sample and then stop", i.e. manual sampling
+	int start(int hz = 100);
+	int reset();
+	~hmc5883();
+
+	int set_range(float max_gauss = 0.88f);
+
+	// ...
+	static device *clone(std::string &path) { return new hmc5883(path); }
+private:
+	// helpers that stash the written value so that we can refresh
+	// it later if necessary
+	int write_CONFIG_A(int val);
+	int write_CONFIG_B(int val);
+	int write_MODE(int val);
+	int config_a, config_b, mode;
+
+	int rate_hz;
+	hmc5883( /* a naked driver makes no sense */ );
+};
+
+static int unused = register_driver("hmc5883", hmc5883::clone);
+
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof(*(a)))
+
+// TODO: operator overloading?
```

Comment:
I like the indexed data approach, but the form here needs to be an abstraction supported by the underlying impl and bus. 

Why is each one defined and opposed to one function? For that matter have an object rev->write()?